include/swift/Demangling/Demangle.h

```diff
@@ -251,6 +251,7 @@ class Node {
   }
 
   static bool deepEquals(const Node *lhs, const Node *rhs) {
+    assert(lhs && rhs && "Nodes being compared may not be null");
```

Comment:
If both pointers can't be null we don't need `if ((!lhs && rhs) || (lhs && !rhs))`. Or we can require at least one of them to be non-null, then the assert should be `lhs || rhs`.

Reply:
Also, if both of them may not be null it might be a good idea to take const references instead

Reply:
Good catch about the now-redundant comparison; removed it.

I think we want to keep the types as pointers, because `NodePointer` (type alias of `Node *`) is the "currency" type used by the demangler; there are no public APIs that consume a `Node &`, so requiring the caller to dereference it in just this case would be a bit unusual.